### ISSUE TITLE
Install Rust tools with `--locked` by default or add a parameter to allow this

### ISSUE THREAD

**Issue by mlegner** [pre-commit/pre-commit#3162]
### search you tried in the issue tracker

rust locked

### describe your actual problem

Many (most?) Rust projects come with lock files that specify the exact versions of dependencies with which they are tested. Sometimes, they don't work with newer versions, which is why Cargo offers the `--locked` parameter to build with the dependency versions specified in the lock files.

Currently, pre-commit installs Rust tools without the `--locked` parameter. I suggest to use this parameter by default or offer a way to specify that a hook should be built using that parameter.

### pre-commit --version

pre-commit 3.6.2

**Comment by asottile** [comment]
does cargo not use the lock file if present by default? I was under the impression that `--locked` is only necessary to produce an error if it is missing

**Comment by mlegner** [comment]
> does cargo not use the lock file if present by default?

No, the lock file is [ignored by default](https://doc.rust-lang.org/cargo/commands/cargo-install.html#dealing-with-the-lockfile).

**Comment by asottile** [comment]
hmmmmm I think `additional_dependencies: [--locked]` will work

**Comment by mlegner** [comment]
No, unfortunately this doesn't seem to work as an `@*` is added automatically:

```
An unexpected error has occurred: CalledProcessError: command: ('~/.cargo/bin/cargo', 'add', '--locked@*')
return code: 1
stdout: (none)
stderr:
    error: unexpected argument '--locked@*' found
    
      tip: a similar argument exists: '--locked'
    
    Usage: cargo add [OPTIONS] <DEP>[@<VERSION>] ...
           cargo add [OPTIONS] --path <PATH> ...
           cargo add [OPTIONS] --git <URL> ...
    
    For more information, try '--help'.
```

**Comment by Timmmm** [comment]
Yeah this is a bit awkward. I have resorted to using `=1.2.3` style dependencies in my `Cargo.toml` file, but this isn't a full solution because it doesn't lock transitive dependencies.

**Comment by bbannier** [comment]
This issue basically makes it impossible for me to package a linter in a pre-commit hook since to build this linter one _has to use locked dependencies_, and there is no officially supported way to pull in pre-built binaries, see #1453.

**Comment by eltos** [comment]
> does cargo not use the lock file if present by default? I was under the impression that --locked is only necessary to produce an error if it is missing

If you @asottile were under this impression, then why not add `--locked` by default now? Why was #3163 rejected?
